File: vermyth/engine/composition.py
import json
import math
from pathlib import Path

from vermyth.contracts import CompositionContract
from vermyth.engine.keys import canonical_aspect_key
from vermyth.schema import (
    Aspect,
    AspectID,
    ContradictionSeverity,
    EffectClass,
    FluidSigil,
    Polarity,
    SemanticVector,
    Sigil,
)


class CompositionEngine(CompositionContract):
    """Core engine: composition of AspectID sets into resolved Sigils from JSON tables."""

    def __init__(
        self,
        data_dir: Path | None = None,
        *,
        contradictions: dict[str, dict] | None = None,
    ) -> None:
        self._data_dir = (
            data_dir
            if data_dir is not None
            else Path(__file__).resolve().parent.parent / "data" / "sigils"
        )
        self._table: dict[str, dict] = {}
        self._contradictions: dict[str, dict] = {}
        self._load_table()
        if contradictions is None:
            self._load_contradictions()
        else:
            self._contradictions = contradictions

    @property
    def contradictions(self) -> dict[str, dict]:
        return self._contradictions

    def _load_contradictions(self) -> None:
        path = self._data_dir / "contradictions.json"
        with path.open(encoding="utf-8") as f:
            loaded = json.load(f)
        if not isinstance(loaded, dict):
            raise ValueError("contradictions.json must contain a JSON object")
        self._contradictions = loaded

    def _ingest_sigil_file(self, path: Path) -> None:
        with path.open(encoding="utf-8") as f:
            entries = json.load(f)
        if not isinstance(entries, list):
            raise ValueError(f"expected JSON array in {path.name}")
        for entry in entries:
            if not isinstance(entry, dict):
                raise ValueError(f"invalid entry in {path.name}")
            raw_names = entry.get("aspects")
            if not isinstance(raw_names, list):
                raise ValueError(f"entry missing aspects array in {path.name}")
            aspect_set: set[AspectID] = set()
            for name in raw_names:
                if not isinstance(name, str):
                    raise ValueError(f"invalid aspect name in {path.name}")
                try:
                    aspect_set.add(AspectID[name])
                except KeyError as exc:
                    raise ValueError(f"unknown AspectID name: {name!r}") from exc
            aspects_fs = frozenset(aspect_set)
            key = canonical_aspect_key(aspects_fs)
            if key in self._table:
                raise ValueError(f"duplicate canonical sigil key: {key!r}")
            self._table[key] = entry

    def _load_table(self) -> None:
        for path in sorted(self._data_dir.glob("*.json")):
            if path.name == "contradictions.json":
                continue
            self._ingest_sigil_file(path)
        extended = self._data_dir / "extended"
        if extended.is_dir():
            for path in sorted(extended.glob("*.json")):
                self._ingest_sigil_file(path)

    def _canonical_key(self, aspects: frozenset[Aspect]) -> str:
        return canonical_aspect_key(aspects)

    def _contradiction_severity_for_key(self, key: str, raw: dict) -> ContradictionSeverity:
        sev_override = raw.get("contradiction_severity")
        if (
            isinstance(sev_override, str)
            and sev_override in ContradictionSeverity.__members__
        ):
            return ContradictionSeverity[sev_override]
        contra = self._contradictions.get(key)
        if contra is None:
            return ContradictionSeverity.NONE
        sev = contra.get("severity", "NONE")
        return ContradictionSeverity[sev]

    def _normalize_unit(self, vec: SemanticVector) -> SemanticVector:
        s = 0.0
        for c in vec.components:
            s += float(c) * float(c)
        n = math.sqrt(s)
        if n == 0.0 or not math.isfinite(n):
            return SemanticVector(
                components=tuple(0.0 for _ in vec.components),
                basis_version=vec.basis_version,
            )
        return SemanticVector(
            components=tuple(float(c) / n for c in vec.components),
            basis_version=vec.basis_version,
        )

    def compose(self, aspects: frozenset[Aspect]) -> Sigil:
        n = len(aspects)
        if n < 1:
            raise ValueError("aspects must contain at least one AspectID")
        if n > 3:
            raise ValueError("aspects must contain at most three AspectIDs")
        key = self._canonical_key(aspects)
        if key not in self._table:
            raise ValueError(f"no defined resolution for aspect combination: {key!r}")
        raw = self._table[key]
        contradiction_severity = self._contradiction_severity_for_key(key, raw)
        name = raw["name"]
        effect_class = EffectClass[raw["effect_class"]]
        resonance_ceiling = float(raw["resonance_ceiling"])
        return Sigil(
            name=name,
            aspects=aspects,
            effect_class=effect_class,
            resonance_ceiling=resonance_ceiling,
            contradiction_severity=contradiction_severity,
        )

    def interpolate(self, vector: SemanticVector, k: int = 3) -> FluidSigil:
        if k < 1:
            raise ValueError("k must be >= 1")

        vec = self._normalize_unit(vector)
        scored: list[tuple[float, str, dict, frozenset[AspectID]]] = []
        for key, raw in self._table.items():
            raw_names = raw.get("aspects")
            if not isinstance(raw_names, list):
                continue
            aspects_set: set[AspectID] = set()
            ok = True
            for name in raw_names:
                if not isinstance(name, str) or name not in AspectID.__members__:
                    ok = False
                    break
                aspects_set.add(AspectID[name])
            if not ok or not aspects_set:
                continue
            aspects_fs = frozenset(aspects_set)
            anchor_vec = SemanticVector.from_aspects(aspects_fs)
            sim = anchor_vec.cosine_similarity(vec)
            scored.append((sim, key, raw, aspects_fs))

        if not scored:
            raise ValueError("no sigil anchors available for interpolation")

        scored.sort(key=lambda t: t[0], reverse=True)
        neighbors = scored[: min(k, len(scored))]

        # Softmax weights over similarities for stability.
        sims = [n[0] for n in neighbors]
        max_sim = max(sims)
        exps = [math.exp(s - max_sim) for s in sims]
        denom = sum(exps) or 1.0
        weights = [e / denom for e in exps]

        weight_by_name: dict[str, float] = {}
        union_aspects: set[AspectID] = set()
        resonance_ceiling = 0.0
        worst_severity = ContradictionSeverity.NONE

        best_idx = max(range(len(neighbors)), key=lambda i: weights[i])
        best_raw = neighbors[best_idx][2]
        best_aspects = neighbors[best_idx][3]
        best_name = str(best_raw.get("name", "UNKNOWN"))
        best_effect_class = EffectClass[best_raw["effect_class"]]

        severity_rank = {
            ContradictionSeverity.NONE: 0,
            ContradictionSeverity.SOFT: 1,
            ContradictionSeverity.HARD: 2,
        }

        for (w, (_sim, key, raw, aspects_fs)) in zip(weights, neighbors, strict=False):
            name = str(raw.get("name", key))
            weight_by_name[name] = weight_by_name.get(name, 0.0) + float(w)
            union_aspects.update(aspects_fs)
            resonance_ceiling += float(raw["resonance_ceiling"]) * float(w)

            if w > 0.10:
                sev = self._contradiction_severity_for_key(key, raw)
                if severity_rank[sev] > severity_rank[worst_severity]:
                    worst_severity = sev

        # Keep aspect pattern compatible with existing seed/registry contracts (1..3).
        if len(union_aspects) > 3:
            union_aspects = set(best_aspects)

        # Polarity from sign balance of the vector components.
        net = 0.0
        for c in vec.components:
            if c > 1e-12:
                net += 1.0
            elif c < -1e-12:
                net -= 1.0
        if net > 0:
            polarity = Polarity.POSITIVE
        elif net < 0:
            polarity = Polarity.NEGATIVE
        else:
            polarity = Polarity.NEUTRAL

        nearest_canonical = best_name
        fluid_name = f"Fluid:{nearest_canonical}"

        # Fingerprint mixes union aspects + vector components for stability.
        fp_basis = (
            "+".join(sorted(a.name for a in union_aspects))
            + "|"
            + ",".join(f"{float(x):.6f}" for x in vec.components[:6])
        )
        semantic_fingerprint = json.dumps(fp_basis).encode("utf-8")
        # Use sha256 via SemanticVector fingerprint style (keep local).
        import hashlib

        fp = hashlib.sha256(semantic_fingerprint).hexdigest()

        return FluidSigil(
            name=fluid_name,
            aspects=frozenset(union_aspects),
            effect_class=best_effect_class,
            resonance_ceiling=float(resonance_ceiling),
            contradiction_severity=worst_severity,
            semantic_fingerprint=fp,
            semantic_vector=vec,
            polarity=polarity,
            source_vector=vector,
            nearest_canonical=nearest_canonical,
            interpolation_weights=weight_by_name,
        )

    def register_sigil_entry(
        self, aspects: frozenset[Aspect], entry: dict, *, allow_override: bool = False
    ) -> None:
        key = self._canonical_key(aspects)
        if key in self._table and not allow_override:
            raise ValueError(
                f"sigil for aspect combination {key!r} already exists; set allow_override=True to replace"
            )
        self._table[key] = entry

    def load_registered_sigils(self, sigils: list[dict]) -> None:
        for row in sigils:
            aspects = row.get("aspects")
            if not isinstance(aspects, frozenset):
                continue
            entry = row.get("entry")
            allow_override = bool(row.get("allow_override", False))
            if not isinstance(entry, dict):
                continue
            self.register_sigil_entry(aspects, entry, allow_override=allow_override)

File: typescript/src/engine/composition.ts
import { createHash } from "node:crypto";
import { readdirSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";

import { sigilsDir } from "../paths.js";
import type { Aspect } from "../schema/aspect.js";
import { aspectIdFromName } from "../schema/aspect.js";
import type { EffectClass } from "../schema/enums.js";
import { ContradictionSeverity as CS, EffectClass as EC } from "../schema/enums.js";
import {
  buildFluidSigil,
  buildSigil,
  fluidPolarityFromVector,
  normalizeSemanticVectorForComposition,
} from "../schema/sigil.js";
import type { SemanticVector } from "../schema/vectors.js";
import { cosineSimilarity, semanticVectorFromAspects } from "../schema/vectors.js";
import { canonicalAspectKey } from "./keys.js";

function parseEffectClass(raw: unknown): EffectClass {
  const s = String(raw);
  if (s in EC) {
    return EC[s as keyof typeof EC];
  }
  throw new Error(`unknown effect_class: ${JSON.stringify(raw)}`);
}

type SigilTableEntry = {
  name: string;
  aspects: string[];
  effect_class: string;
  resonance_ceiling: number;
  contradiction_severity?: string;
};

export class CompositionEngine {
  private readonly dataDir: string;
  private readonly table: Map<string, Record<string, unknown>> = new Map();
  private contradictions: Record<string, Record<string, unknown>> = {};

  constructor(
    dataDir?: string,
    contradictions?: Record<string, Record<string, unknown>> | null,
  ) {
    this.dataDir = dataDir ?? sigilsDir();
    this.loadTable();
    if (contradictions === undefined || contradictions === null) {
      this.loadContradictions();
    } else {
      this.contradictions = contradictions;
    }
  }

  get contradictionsMap(): Record<string, Record<string, unknown>> {
    return this.contradictions;
  }

  private loadContradictions(): void {
    const path = join(this.dataDir, "contradictions.json");
    try {
      const raw = readFileSync(path, "utf8");
      const loaded = JSON.parse(raw) as unknown;
      if (loaded && typeof loaded === "object" && !Array.isArray(loaded)) {
        this.contradictions = loaded as Record<string, Record<string, unknown>>;
      } else {
        this.contradictions = {};
      }
    } catch {
      this.contradictions = {};
    }
  }

  private ingestSigilFile(path: string): void {
    const entries = JSON.parse(readFileSync(path, "utf8")) as unknown;
    if (!Array.isArray(entries)) {
      throw new Error(`expected JSON array in ${path}`);
    }
    for (const entry of entries) {
      if (!entry || typeof entry !== "object") {
        throw new Error(`invalid entry in ${path}`);
      }
      const e = entry as SigilTableEntry;
      const rawNames = e.aspects;
      if (!Array.isArray(rawNames)) {
        throw new Error(`entry missing aspects array in ${path}`);
      }
      const aspectSet = new Set<Aspect>();
      for (const name of rawNames) {
        if (typeof name !== "string") {
          throw new Error(`invalid aspect name in ${path}`);
        }
        aspectSet.add(aspectIdFromName(name));
      }
      const aspectsFs = aspectSet;
      const key = canonicalAspectKey(aspectsFs);
      if (this.table.has(key)) {
        throw new Error(`duplicate canonical sigil key: ${JSON.stringify(key)}`);
      }
      this.table.set(key, entry as Record<string, unknown>);
    }
  }

  private loadTable(): void {
    const dir = this.dataDir;
    for (const name of readdirSync(dir).sort()) {
      if (!name.endsWith(".json")) {
        continue;
      }
      if (name === "contradictions.json") {
        continue;
      }
      this.ingestSigilFile(join(dir, name));
    }
    const extended = join(dir, "extended");
    try {
      if (statSync(extended).isDirectory()) {
        for (const name of readdirSync(extended).sort()) {
          if (name.endsWith(".json")) {
            this.ingestSigilFile(join(extended, name));
          }
        }
      }
    } catch {
      /* optional */
    }
  }

  private contradictionSeverityForKey(key: string, raw: Record<string, unknown>): CS {
    const sevOverride = raw["contradiction_severity"];
    if (sevOverride === "NONE" || sevOverride === "SOFT" || sevOverride === "HARD") {
      return sevOverride;
    }
    const contra = this.contradictions[key];
    if (!contra) {
      return CS.NONE;
    }
    const sev = contra["severity"];
    if (sev === "NONE" || sev === "SOFT" || sev === "HARD") {
      return sev;
    }
    return CS.NONE;
  }

  compose(aspects: ReadonlySet<Aspect>): ReturnType<typeof buildSigil> {
    const n = aspects.size;
    if (n < 1) {
      throw new Error("aspects must contain at least one AspectID");
    }
    if (n > 3) {
      throw new Error("aspects must contain at most three AspectIDs");
    }
    const key = canonicalAspectKey(aspects);
    const raw = this.table.get(key);
    if (!raw) {
      throw new Error(`no defined resolution for aspect combination: ${JSON.stringify(key)}`);
    }
    const contradictionSeverity = this.contradictionSeverityForKey(key, raw);
    const name = String(raw["name"]);
    const effectClass = parseEffectClass(raw["effect_class"]);
    const resonanceCeiling = Number(raw["resonance_ceiling"]);
    return buildSigil({
      name,
      aspects,
      effectClass,
      resonanceCeiling,
      contradictionSeverity,
    });
  }

  interpolate(vector: SemanticVector, k = 3): ReturnType<typeof buildFluidSigil> {
    if (k < 1) {
      throw new Error("k must be >= 1");
    }
    const vec = normalizeSemanticVectorForComposition(vector);
    const scored: Array<{
      sim: number;
      key: string;
      raw: Record<string, unknown>;
      aspects: ReadonlySet<Aspect>;
    }> = [];

    for (const [key, raw] of this.table) {
      const rawNames = raw["aspects"];
      if (!Array.isArray(rawNames)) {
        continue;
      }
      const aspectsSet = new Set<Aspect>();
      let ok = true;
      for (const name of rawNames) {
        if (typeof name !== "string") {
          ok = false;
          break;
        }
        try {
          aspectsSet.add(aspectIdFromName(name));
        } catch {
          ok = false;
          break;
        }
      }
      if (!ok || aspectsSet.size === 0) {
        continue;
      }
      const aspectsFs = aspectsSet;
      const anchorVec = semanticVectorFromAspects(aspectsFs);
      const sim = cosineSimilarity(anchorVec, vec);
      scored.push({ sim, key, raw, aspects: aspectsFs });
    }

    if (scored.length === 0) {
      throw new Error("no sigil anchors available for interpolation");
    }

    scored.sort((a, b) => b.sim - a.sim);
    const neighbors = scored.slice(0, Math.min(k, scored.length));

    const sims = neighbors.map((n) => n.sim);
    const maxSim = Math.max(...sims);
    const exps = sims.map((s) => Math.exp(s - maxSim));
    const denom = exps.reduce((a, b) => a + b, 0) || 1;
    const weights = exps.map((e) => e / denom);

    const weightByName: Record<string, number> = {};
    const unionAspects = new Set<Aspect>();
    let resonanceCeiling = 0;
    let worstSeverity: CS = CS.NONE;

    const severityRank: Record<CS, number> = {
      [CS.NONE]: 0,
      [CS.SOFT]: 1,
      [CS.HARD]: 2,
    };

    let bestIdx = 0;
    for (let i = 1; i < neighbors.length; i++) {
      if (weights[i]! > weights[bestIdx]!) {
        bestIdx = i;
      }
    }
    const bestRaw = neighbors[bestIdx]!.raw;
    const bestAspects = neighbors[bestIdx]!.aspects;
    const bestName = String(bestRaw["name"] ?? "UNKNOWN");
    const bestEffectClass = parseEffectClass(bestRaw["effect_class"]);

    for (let i = 0; i < neighbors.length; i++) {
      const w = weights[i]!;
      const { key, raw, aspects: aspectsFs } = neighbors[i]!;
      const name = String(raw["name"] ?? key);
      weightByName[name] = (weightByName[name] ?? 0) + w;
      for (const a of aspectsFs) {
        unionAspects.add(a);
      }
      resonanceCeiling += Number(raw["resonance_ceiling"]) * w;
      if (w > 0.1) {
        const sev = this.contradictionSeverityForKey(key, raw);
        if (severityRank[sev] > severityRank[worstSeverity]) {
          worstSeverity = sev;
        }
      }
    }

    let union = unionAspects;
    if (union.size > 3) {
      union = new Set(bestAspects);
    }

    const polarity = fluidPolarityFromVector(vec);
    const nearestCanonical = bestName;
    const fluidName = `Fluid:${nearestCanonical}`;

    const fpBasis =
      [...union]
        .map((a) => a.name)
        .sort()
        .join("+") +
      "|" +
      [...vec.components.slice(0, 6)].map((x) => `${Number(x).toFixed(6)}`).join(",");
    const fp = createHash("sha256").update(JSON.stringify(fpBasis), "utf8").digest("hex");

    return buildFluidSigil({
      name: fluidName,
      aspects: union,
      effectClass: bestEffectClass,
      resonanceCeiling,
      contradictionSeverity: worstSeverity,
      semanticFingerprint: fp,
      semanticVector: vec,
      polarity,
      sourceVector: vector,
      nearestCanonical,
      interpolationWeights: weightByName,
    });
  }

  registerSigilEntry(
    aspects: ReadonlySet<Aspect>,
    entry: Record<string, unknown>,
    opts?: { allowOverride?: boolean },
  ): void {
    const key = canonicalAspectKey(aspects);
    if (this.table.has(key) && !opts?.allowOverride) {
      throw new Error(
        `sigil for aspect combination ${JSON.stringify(key)} already exists; set allowOverride to replace`,
      );
    }
    this.table.set(key, entry);
  }
}
